codegenerator/cli/templates/static/codegen/src/Prometheus.res

```diff
@@ -40,49 +40,131 @@ let sourceChainHeight = PromClient.Gauge.makeGauge({
   "labelNames": ["chainId"],
 })
 
-let benchmarkSummaryData = PromClient.Gauge.makeGauge({
-  "name": "benchmark_summary_data",
-  "help": "All data points collected during indexer benchmark",
-  "labelNames": ["group", "label", "stat"],
-})
+module SafeGauge: {
+  type t<'a>
+  let makeOrThrow: (~name: string, ~help: string, ~labelSchema: S.t<'a>) => t<'a>
+  let setInt: (t<'a>, ~labels: 'a, ~value: int) => unit
+  let setFloat: (t<'a>, ~labels: 'a, ~value: float) => unit
+} = {
+  type t<'a> = PromClient.Gauge.gauge
+
+  let rec schemaIsString = (schema: S.t<'a>) =>
+    switch schema->S.classify {
+    | String => true
+    | Null(s)
+    | Option(s) =>
+      s->schemaIsString
+    | _ => false
+    }
+
+  let getLabelNames = (schema: S.t<'a>) =>
+    switch schema->S.classify {
+    | Object({items}) =>
+      let nonStringFields = items->Belt.Array.reduce([], (nonStringFields, item) => {
+        if item.schema->schemaIsString {
+          nonStringFields
+        } else {
+          nonStringFields->Belt.Array.concat([item.location])
+        }
+      })
+
+      switch nonStringFields {
+      | [] => items->Belt.Array.map(item => item.location)->Ok
+      | nonStringItems =>
+        let nonStringItems = nonStringItems->Js.Array2.joinWith(", ")
+        Error(
+          `Label schema must be an object with string (or optional string) values. Non string values: ${nonStringItems}`,
+        )
+      }
+    | _ => Error("Label schema must be an object")
+    }
+
+  let makeOrThrow = (~name, ~help, ~labelSchema: S.t<'a>): t<'a> =>
+    switch labelSchema->getLabelNames {
+    | Ok(labelNames) =>
+      PromClient.Gauge.makeGauge({
+        "name": name,
+        "help": help,
+        "labelNames": labelNames,
+      })
+
+    | Error(error) => Js.Exn.raiseError(error)
+    }
+
+  let makeLabels = (self: t<'a>, ~labels: 'a): t<'a> => self->PromClient.Gauge.labels(labels)
+
+  let setFloat = (self: t<'a>, ~labels: 'a, ~value) =>
+    self
+    ->makeLabels(~labels)
+    ->PromClient.Gauge.setFloat(value)
+
+  let setInt = (self: t<'a>, ~labels: 'a, ~value) =>
+    self
+    ->makeLabels(~labels)
+    ->PromClient.Gauge.set(value)
+}
```

Comment:
Not completely necessary but thought it's one of the nice things we can do for safety using rescript schema